clang/include/clang/CAS/IncludeTree.h

```diff
@@ -0,0 +1,262 @@
+//===- IncludeTree.h - Include-tree CAS graph -------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_CAS_CASINCLUDETREE_H
+#define LLVM_CLANG_CAS_CASINCLUDETREE_H
+
+#include "clang/Basic/SourceManager.h"
+#include "llvm/CAS/CASDB.h"
+
+namespace llvm {
+class SmallBitVector;
+}
+
+namespace clang {
+namespace cas {
+
+/// Base class for include-tree related nodes. It makes it convenient to
+/// add/skip/check the "node kind identifier" (\p getNodeKind()) that is put
+/// at the beginning of the object data for every include-tree related node.
+template <typename NodeT> class IncludeTreeBase : public ObjectProxy {
```

Comment:
Is there reason not to use `llvm::cas::NodeSchema` to organize all the different object type?

Reply:
(per offline discussion with Steven)

I think the `llvm::cas::NodeSchema` is a bit too heavyweight specifically for the purposes of these handful of include-tree nodes, I could implement the functionality via the proxy classes directly which seemed simpler with less maintenance overhead.